lib/fed/decoder.go

```diff
@@ -30,6 +38,23 @@ func NewDecoder(r io.Reader) *Decoder {
 	return d
 }
 
+func CloneDecoder(d *Decoder, r io.Reader) *Decoder {
```

Comment:
instead of cloning the decoder, i usually use fed.ToConcrete to convert the packet. It does change the decoder though so you have to return the concrete packet if you call that. 

maybe instead we should make pending packet hold an additional field `concrete any` and repeated calls to ToConcrete with the pending packet will just return the concrete version. this way you can just always use fed.ToConcrete[T] everywhere instead of doing type checks and clones and not worry about effecting the underlying decoder

Reply:
something like this in ToConcrete
```go
	case *PendingPacket:
		if p.Concrete != nil {
			*value = p.Concrete.(T)
			return nil
		}

		if err := value.ReadFrom(p.Decoder); err != nil {
			return err
		}
		p.Concrete = *value
		return nil
```

Reply:
I would follow your lead as you have the expertise. As you prolly are aware, I just needed a way to "peek" without disturbing the decoder or it's io.Reader. if the cost of p.Concrete is fairly low then that seems like a good way to go

Reply:
yeah ToConcrete just does what you were doing anyways